distribution/lib/Standard/Base/0.0.0-dev/src/Data/Decimal.enso

```diff
@@ -0,0 +1,582 @@
+     result; results are rounded to this precision.
+   - rounding_mode: the method to use for rounding. See `Rounding_Mode` for
+     details about each rounding method.
+
+   When a `Math_Context` value is used in an operation, and causes a loss of
+   precision, a `Loss_Of_Numeric_Precision` warning is attached to the result.
+
+   If no `Math_Context` value is supplied, the default is to make all operations
+   exact. A `Math_Context` with a precision of `0` has the same effect.
+
+   A `Decimal` can represent any `Float` precisely. However, some `Float`
+   literals, such as `0.1`, cannot be represented exactly to infinite
+   precision by a `Float`. For this reason, constructing a `Decimal` from
+   a `Float` always attaches a `Loss_Of_Numeric_Precision` warning to the
+   result. To avoid this problem, it is recommended to create `Decimal`s from
+   fractional values by passing the value in as a `Text`, where possible and
+   convenient.
+
+   The `Decimal` class provides arithmetic operations `add`, `subtract`,
+   `multiply`, and `divide`, which can take a `Math_Context` argument. You can
+   also use the usual operators `+`, `-`, `*`, `/`, which are the same as the
+   named methods, but which cannot take a `Math_Context`.
+
+   In the case of `divide`, it is possible that the result will have a
+   non-terminating deicmal expansion. If the operation did not specify a
+   `Math_Context`, or specified an explicit `Math_Context` with infinite
+   precision, then it is impossible to represent the result as requested, and an
+   `Arithmetic_Error` will be thrown. In this case, the solution is to specify
+   an explicit precision using a `Math_Context`.
+type Decimal
+    Value (big_decimal:BigDecimal)
+
+    ## Construct a `Decimal` from a string or integer.
+
+       Arguments:
+       - x: The `Text`, `Integer`, or `Float` to construct a `Decimal` from.
+       - mc: The `Math_Context` to use to specify precision and `Rounding_Mode`.
+         If a `Math_Context` is used, there is a possibility of a loss of
+         precision.
+
+       ? Number Format
+
+         The textual format for a Decimal is defined at
+         https://docs.oracle.com/javase/8/docs/api/java/math/BigDecimal.html#BigDecimal-java.lang.String-.
+
+       ! Error Conditions
+
+         - If the `Text` argument is incorrectly formatted, a `Number_Parse_Error`
+           is thrown.
+         - If the construction of the Decimal results in a loss of precision, a
+           `Loss_Of_Numeric_Precision` warning is attached. This can only happen
+           if a `Math_Context` value is explicitly passed.
+
+        ^ Example
+          Create a `Decimal` from a string.
+
+              c = Decimal.new "12.345"
+    new : Text | Integer | Float -> Math_Context -> Decimal ! Arithmetic_Error | Number_Parse_Error
+    new (x : Text | Integer | Float) (mc : Math_Context | Nothing)=Nothing -> Decimal | Number_Parse_Error =
```

Comment:
Just like in Java, by default we use exact arithmetic for all operations. This might turn out to be confusing to users who simply want to construct 1/3 and find that it gives them an error. But I think it's the right choice for a first pass, since picking a non-exact default isn't straightforward.